accounts/models.py

```diff
@@ -350,55 +381,123 @@ def num_sounds_pending_moderation(self):
+        account).
 
-        This method anonymise the user and flags it as deleted. If
-        remove_sounds is True then the Sound (and Pack) object is removed from
-        the database.
+        Optionally, a user can be fully deleted from DB including all of its packs, sounds and other relaed content.
+        Even in this case a DeletedUser object will be created to keep a record.
+
+        Args:
+            remove_sounds (bool): if True the sounds created by the user will be deleted as well. Otherwise the sounds
+              will still be available to other users but appear under a "deleted user" account. Defaults to False.
+            delete_user_object_from_db (bool): if True the user object will be completely removed from the DB together
+              with all related content. Defaults to False.
+            deletion_reason (str): reason for the user being deleted. Should be one of the choices defined in
+              DeletedUser.DELETION_REASON_CHOICES. Defaults to DeletedUser.DELETION_REASON_DELETED_BY_ADMIN.
         """
 
-        self.user.username = 'deleted_user_%s' % self.user.id
-        self.user.email = 'deleted_user_%s@freesound.org' % self.user.id
-        self.has_avatar = False
-        self.is_deleted_user = True
-        self.user.set_unusable_password()
-
-        self.about = ''
-        self.home_page = ''
-        self.signature = ''
-        self.geotag = None
-
-        self.save()
-        self.user.save()
-        if remove_sounds:
-            Sound.objects.filter(user=self.user).delete()
-            Pack.objects.filter(user=self.user).update(is_deleted=True)
-        else:
-            Sound.objects.filter(user=self.user).update(is_index_dirty=True)
+        # Run all deletion operations in a single transaction so if there is an error we don't create duplicate
+        # DeletedUser objects
+        with transaction.atomic():
+
+            # Create a DeletedUser object to store basic information for the record (first check if it
+            # already exists because user was deleted previously but db object preserved
+            try:
+                deleted_user_object = DeletedUser.objects.get(user=self.user)
+            except DeletedUser.DoesNotExist:
+                deleted_user_object = DeletedUser.objects.create(
+                    user=self.user,
+                    username=self.user.username,
+                    email=self.user.email,
+                    date_joined=self.user.date_joined,
+                    last_login=self.user.last_login,
+                    reason=deletion_reason)
+
+            # If UserGDPRDeletionRequest object(s) exist for that user, update their status and deleted_user property
+            UserGDPRDeletionRequest.objects.filter(user_id=self.user.id)\
+                .update(status="de", deleted_user=deleted_user_object)
+
+            # If Sound/Pack objects will be deleted, we need to make sure that num_sound_downloads and
```

Comment:
This bit could be simplified if we somehow removed the >=0 constrain in the count fields...

Reply:
Addressed in c65491773db20f10928cd6c2b7288f0f52292c86. Now we use PG's `Greatest` to make sure the count fields are not set to < 0 and therefore we avoid the need of updating the fields before deletion.